### PR TITLE
Remove macro expansion and replace uses with FE typing + BE lowering

### PR DESCRIPTION
**Pull request by gmarkall**
As title. WIP PR to test on CI, and for testing the change to the ROC backend (I don't have a device to test). This will need a bit of tidying up and refactoring before I remove the WIP tag.

Also includes the fix to #5408, for testing convenience.

@stuartarchibald are you able to try this on a ROC device please?

### CODE REVIEW

**Comment by gmarkall** [comment]
Also, it looks like there's no use for the `ir.Intrinsic` node anymore - https://github.com/numba/numba/blob/master/numba/core/ir.py#L1025 - it seemed to only be used in macro expansion. There is `numba.core.extending._Intrinsic` ( https://github.com/numba/numba/blob/master/numba/core/extending.py#L258 ) but I think it is separate / different.

Any thoughts on the removal of `ir.Intrinsic` along with the macro expansion?

**Comment by esc** [comment]
@gmarkall thanks for submitting this, I have marked it as in-progress for now.

**Comment by gmarkall** [comment]
@esc Thanks - I'm going to split it in two as Siu suggested so that the CUDA changes can go in without needing to be blocked on testing the one change to the ROC target. I'll keep this PR for the removal of all macro expansion code, and make a new one with just CUDA.

**Comment by gmarkall** [comment]
A CUDA-only version of this PR is in #5481, which is ready for review - once that is approved and/or merged, I will amend this PR accordingly.

**Comment by gmarkall** [comment]
Now that #5481 is merged, I've merged master into this branch and updated this PR, and I believe it is now ready for review. It now consists of:

* Removal of the macro expansion pass and associated classes.
* Removal of the `Intrinsic` IR node, which is no longer used. "Intrinsic"  is still a bit of an overloaded term - there's the IR intrinsic nodes, "intrinsic" calls like `getitem` in typing (see e.g. uses of `numba.core.typing.TypeInferer.typeof_intrinsic_call`), and `numba.core.extending._Intrinsic`.
* Some clarifications to comments of CUDA tests, which got missed from #5481.
* Modification of the ROC/HSA target to swap the use of a macro for typing and lowering. This is done blind without testing - if someone with hardware access could test this and let me know if it works, that would be much appreciated! I do have one question about my changes, which I'll add to the diff.

File: numba/roc/hsaimpl.py
Comment (gmarkall) [inline numba/roc/hsaimpl.py:238]: I can't work out why there was never a lowering for `hsail.smem.alloc` with an integer argument. I've swapped like-for-like here, but I can't tell whether this is going to work. Does another lowering:

```python
@lower(roc.shared.array, types.IntegerLiteral, types.Any)
```

need adding, or is there something I've missed that always converts these into `UniTuple`s?

Secondly, if there are two integer literals of different value, I suspect the argument type might be `types.Tuple` - does this case need handling?

Comment (sklam) [reply numba/roc/hsaimpl.py:238]: I can't tell why this works neither. We will need to check this backend on real HW.

Comment (sklam) [reply numba/roc/hsaimpl.py:238]: And for `Tuple` vs `UniTuple`, you are right:

```
In [3]: types.UniTuple.from_types([types.IntegerLiteral(1), types.IntegerLiteral(2)])
Out[3]: Tuple(Literal[int](1), Literal[int](2))

In [4]: types.UniTuple.from_types([types.IntegerLiteral(1), types.IntegerLiteral(1)])
Out[4]: UniTuple(Literal[int](1) x 2)
```

will need a version for `Tuple`

Comment (stuartarchibald) [reply numba/roc/hsaimpl.py:238]: I'll try and get some hardware sorted locally to test this on.

**Comment by gmarkall** [comment]
Per recent discussions (in the dev meeting, IIRC) it's not urgent or likely to get this into 0.51, so I've bumped the milestone to 0.52.

**Comment by stuartarchibald** [comment]
> Per recent discussions (in the dev meeting, IIRC) it's not urgent or likely to get this into 0.51, so I've bumped the milestone to 0.52.

Thanks @gmarkall

File: numba/roc/hsaimpl.py
Comment (stuartarchibald) [inline numba/roc/hsaimpl.py:238]: This needs `from numba import roc` at the top.

File: numba/roc/hsaimpl.py
Comment (gmarkall) [reply numba/roc/hsaimpl.py:248]: Hopefully this is now resolved.

File: numba/roc/hsaimpl.py
Comment (gmarkall) [reply numba/roc/hsaimpl.py:239]: Added.

File: numba/cuda/tests/cudapy/test_lang.py
Comment (stuartarchibald) [inline numba/cuda/tests/cudapy/test_lang.py:44]:
```suggestion
        more than one block. Was originally to ensure that macro expansion works
```

**Comment by stuartarchibald** [comment]
For 5daa538 the things I'd expect to work ok on ROCm in light of these changes still work, so I think it's ok on that hardware target. The patch itself looks good to me, @sklam did you want to take another look? If not, this can go through the build farm and be merged if it passes. Thanks again for the large refactoring effort, very pleased to see `Macro` removed!

**Comment by sklam** [comment]
It will just need to be tested on real ROC hw i the farm to pick up anything else we missed by eyeballing.

**Comment by stuartarchibald** [comment]
> It will just need to be tested on real ROC hw in the farm to pick up anything else we missed by eyeballing.

I've run this on real ROC hw locally, the things I'd expect to work ok are working.

**Review by stuartarchibald — APPROVED** [review]
Thanks for working on this and for all the fixes etc, especially guessing for non-local hardware! Looks good!

**Comment by esc** [comment]
BFID: numba_smoketest_cuda_98

**Comment by esc** [comment]
Buildfarm passed.

**Pull request merged**
